src/buffer.rs

```diff
@@ -357,3 +368,66 @@ pub enum EncMacroblockParameterBuffer {
     /// Abstraction over `VAEncMacroblockParameterBufferH264`
     H264(h264::EncMacroblockParameterBufferH264),
 }
+
+/// Helper to access a single segment of mapped coded buffer
+pub struct CodedBufferSegment<'s> {
+    pub bit_offset: u32,
+    pub status: u32,
+    pub buf: &'s [u8],
+}
+
+/// Helper to access segments of mapped coded buffer
+pub struct CodedBuffer<'p> {
+    segments: Vec<CodedBufferSegment<'p>>,
+    buffer: &'p Buffer,
+}
+
+impl<'p> CodedBuffer<'p> {
+    /// Map a 'VAEncCodedBufferType' buffer. `picture` makes sure that the surface was synced and
+    /// would not change state until this buffer is unmapped (dropped).
+    pub fn new(buffer: &'p Buffer) -> Result<Self, VaError> {
```

Comment:
I am OK with this PR in general. Can we only assert that we have the right type of buffer here? As I said, this can map any `Buffer` as `CodedBuffer`, which is a dangerous misuse of the API.

Reply:
Yeah, this is problematic. I'd have to add a buffer type to `Buffer` structure. Currently it does not hold this information at all. If you're fine with that I'll add it.

Reply:
Maybe a simple bool will do? We can set it in `Buffer::new` if we match here:

`BufferType::EncCodedBuffer(size) => (std::ptr::null_mut(), size),`

This way we don't have to carry `BufferType` around.

Reply:
Maybe we can use the type system for this.

The idea would be to turn `CodedBuffer` into something like this:

```
pub struct EncCodedBuffer(Buffer);
```

With an accessor method to get a reference to the wrapper `Buffer`. We don't make it `pub` because that would allow anyone to create an object of this type, which we want to avoid because an `EncCodedBuffer` can only be created by the `new_enc_coded` method of `Buffer`, which basically calls `new` with the `BufferType::EncCodedBuffer` buffer type and wraps the result.

Then `CodedBuffer::new` would take an argument of type `&'p EncCodedBuffer`, and that way we know it cannot be used on anything else than buffers of the right type.

Does that sound reasonable?

Reply:
I like this idea. I modified the code accordingly. Also I renamed the `CodedBuffer` and `CodedBufferSegment` to `MappedCodedBuffer` and `MappedCodedSegment` to better reflect their purpose. PTAL and please let me know what do you think @dwlsalmeida @Gnurou.